src/lavinmq/mqtt/client.cr

```diff
@@ -0,0 +1,261 @@
+            @log.debug { "Recieved disconnect" }
+            break
+          end
+        end
+      rescue ex : ::MQTT::Protocol::Error::PacketDecode
+        @log.warn(exception: ex) { "Packet decode error" }
+        publish_will
+      rescue ex : MQTT::Error::Connect
+        @log.warn { "Connect error: #{ex.message}" }
+      rescue ex : ::IO::Error
+        @log.error { "Client unexpectedly closed connection: #{ex.message}" } unless @closed
+        publish_will
+      rescue ex
+        @log.error(exception: ex) { "Read Loop error" }
+        publish_will
+      ensure
+        @broker.remove_client(self)
+        @waitgroup.done
+        close_socket
+      end
+
+      def read_and_handle_packet
+        packet : MQTT::Packet = MQTT::Packet.from_io(@io)
+        @log.trace { "Recieved packet:  #{packet.inspect}" }
+        @recv_oct_count += packet.bytesize
+
+        case packet
+        when MQTT::Publish     then recieve_publish(packet)
+        when MQTT::PubAck      then recieve_puback(packet)
+        when MQTT::Subscribe   then recieve_subscribe(packet)
+        when MQTT::Unsubscribe then recieve_unsubscribe(packet)
+        when MQTT::PingReq     then receive_pingreq(packet)
+        when MQTT::Disconnect  then return packet
+        else                        raise "received unexpected packet: #{packet}"
+        end
+        packet
+      end
+
+      def send(packet)
+        @lock.synchronize do
+          packet.to_io(@io)
+          @socket.flush
+        end
+        @send_oct_count += packet.bytesize
+      end
+
+      def receive_pingreq(packet : MQTT::PingReq)
+        send MQTT::PingResp.new
+      end
+
+      def recieve_publish(packet : MQTT::Publish)
+        @broker.publish(packet)
+        # Ok to not send anything if qos = 0 (fire and forget)
+        if packet.qos > 0 && (packet_id = packet.packet_id)
+          send(MQTT::PubAck.new(packet_id))
+        end
+      end
+
+      def recieve_puback(packet : MQTT::PubAck)
+        @broker.sessions[@client_id].ack(packet)
+      end
+
+      def recieve_subscribe(packet : MQTT::Subscribe)
+        qos = @broker.subscribe(self, packet)
+        send(MQTT::SubAck.new(qos, packet.packet_id))
+      end
+
+      def recieve_unsubscribe(packet : MQTT::Unsubscribe)
+        @broker.unsubscribe(self, packet)
+        send(MQTT::UnsubAck.new(packet.packet_id))
+      end
+
+      def details_tuple
+        {
+          vhost:        @broker.vhost.name,
+          user:         @user.name,
+          protocol:     "MQTT",
+          client_id:    @client_id,
+          connected_at: @connected_at,
+        }.merge(stats_details)
+      end
+
+      private def publish_will
+        return unless will = @will
+        packet = MQTT::Publish.new(
+          topic: will.topic,
+          payload: will.payload,
+          packet_id: nil,
+          qos: will.qos,
+          retain: will.retain?,
+          dup: false,
+        )
+        @broker.publish(packet)
```

Comment:
return yearly or if assigments, i think i prefer if assignments, not unless assignments. 

```suggestion
        if will = @will
          @broker.publish MQTT::Publish.new(
            topic: will.topic,
            payload: will.payload,
            packet_id: nil,
            qos: will.qos,
            retain: will.retain?,
            dup: false,
          )
        end
```